File: src/main/java/cs/android/viewbase/CSMenuItem.java
package cs.android.viewbase;

import cs.java.callback.CSRun;
import cs.java.callback.CSRunWith;

import static cs.java.lang.CSLang.NO;
import static cs.java.lang.CSLang.YES;
import static cs.java.lang.CSLang.is;

/**
 * Created by Rene Dohan on 18/1/17.
 */
public class CSMenuItem {
    private CSViewController _controller;
    private int _id;
    private CSRun _run;
    private boolean _visible = YES;
    private CSRunWith<Boolean> _runCheckable;
    private boolean _isChecked;

    CSMenuItem(CSViewController controller, int id) {
        _controller = controller;
        _id = id;
    }

    public CSMenuItem onClick(CSRun run) {
        _run = run;
        return this;
    }

    public CSMenuItem onClick(CSRunWith<Boolean> run) {
        _runCheckable = run;
        return this;
    }

    public int id() {
        return _id;
    }

    public void run() {
        if (is(_run)) _run.run();
    }

    public void run(boolean isChecked) {
        if (is(_runCheckable)) _runCheckable.run(isChecked);
    }

    public CSMenuItem hide() {
        _visible = NO;
        if (_controller.isCreated()) _controller.invalidateOptionsMenu();
        return this;
    }

    public boolean isVisible() {
        return _visible;
    }

    public void show() {
        _visible = YES;
        if (_controller.isCreated()) _controller.invalidateOptionsMenu();
    }

    public CSMenuItem visible(boolean visible) {
        _visible = visible;
        if (_controller.isCreated()) _controller.invalidateOptionsMenu();
        return this;
    }

    public void checked(boolean checked) {
        _isChecked = checked;
    }

    public boolean isChecked() {
        return _isChecked;
    }

    public void onChecked(CSOnMenuItem onItem) {
        checked(!isChecked());
        onItem.checked(isChecked());
        run(isChecked());
    }
}

File: library/renetik-controller/src/main/java/menu/CSMenuItem.kt
package renetik.android.controller.menu

import android.view.MenuItem.*
import android.view.View
import renetik.android.controller.base.CSViewController

/**
 * Created by Rene Dohan on 18/1/17.
 */
open class CSMenuItem {

    companion object {
        private var lastGeneratedMenuItemId = 10
    }

    val id: Int
    val isGenerated: Boolean
    private val controller: CSViewController<*>
    var title: String? = null
    var isVisible = true
    var isChecked: Boolean? = null
    var iconResource = 0
    var showAsAction = SHOW_AS_ACTION_IF_ROOM
    private var runWith: ((CSMenuItem) -> Unit)? = null
    var actionView: View? = null

    constructor(controller: CSViewController<*>, id: Int) {
        this.controller = controller
        this.id = id
        isGenerated = false
    }

    constructor(controller: CSViewController<*>, title: String) {
        this.controller = controller
        this.id = lastGeneratedMenuItemId++
        isGenerated = true
        this.title = title
    }

    fun onClick(run: (CSMenuItem) -> Unit) = apply { runWith = run }
    fun run() = runWith?.invoke(this)
    fun hide() = visible(false)
    fun show() = visible(true)

    fun visible(visible: Boolean) = apply {
        if (isVisible == visible) return this
        isVisible = visible
        if (controller.isCreated) controller.invalidateOptionsMenu()
    }

    fun onChecked(onItem: CSOnMenuItem) {
        isChecked = !isChecked!!
        onItem.isChecked = isChecked!!
        run()
    }

    fun setIconResourceId(iconResourceId: Int) = apply { this.iconResource = iconResourceId }
    fun neverAsAction() = apply { showAsAction = SHOW_AS_ACTION_NEVER }
    fun alwaysAsAction() = apply { showAsAction = SHOW_AS_ACTION_ALWAYS }
    fun withText() = apply { showAsAction = showAsAction or SHOW_AS_ACTION_WITH_TEXT }
    fun remove() = controller.removeMenuItem(this)
}

